src/scene/lightmapper.js

```diff
@@ -244,641 +491,526 @@ class Lightmapper {
-        var dilate = shaderChunks.dilatePS;
+        for (let i = 0; i < sceneLights.length; i++) {
+            let light = sceneLights[i];
 
-        var dilateShader = createShaderFromCode(device, shaderChunks.fullscreenQuadVS, dilate, "lmDilate");
-        var constantTexSource = device.scope.resolve("source");
-        var constantPixelOffset = device.scope.resolve("pixelOffset");
-        var constantBakeDir = device.scope.resolve("bakeDir");
+            // store all lights and their original settings we need to temporariy modify
+            const lightInfo = new LightInfo(light);
+            allLights.push(lightInfo);
 
-        var pixelOffset = new Float32Array(2);
+            // bake light
+            if (light.enabled && (light.mask & MASK_LIGHTMAP) !== 0) {
 
-        var drawCalls = activeComp._meshInstances;
+                // if baked, it can't be used as static
+                light.isStatic = false;
 
-        // update scene matrices
-        for (i = 0; i < drawCalls.length; i++) {
-            if (drawCalls[i].node) drawCalls[i].node.getWorldTransform();
+                light.mask = 0xFFFFFFFF;
+                light.shadowUpdateMode = light.type === LIGHTTYPE_DIRECTIONAL ? SHADOWUPDATE_REALTIME : SHADOWUPDATE_THISFRAME;
+                bakeLights.push(lightInfo);
+            }
         }
 
-        // Store scene values
-        var origFog = scene.fog;
-        var origAmbientR = scene.ambientLight.r;
-        var origAmbientG = scene.ambientLight.g;
-        var origAmbientB = scene.ambientLight.b;
+        // sort bake lights by type to minimize shader switches
+        bakeLights.sort();
+    }
 
-        scene.fog = FOG_NONE;
-        scene.ambientLight.set(0, 0, 0);
+    restoreLights(allLights) {
 
-        // Create pseudo-camera
-        if (!lmCamera) {
-            lmCamera = new Camera();
-            lmCamera.clearColor = new Color(0, 0, 0, 0);
-            lmCamera.clearColorBuffer = true;
-            lmCamera.clearDepthBuffer = false;
-            lmCamera.clearStencilBuffer = false;
-            lmCamera.frustumCulling = false;
-            lmCamera.node = new GraphNode();
+        for (let i = 0; i < allLights.length; i++) {
+            allLights[i].restore();
         }
+    }
 
-        var node;
-        var lm, rcv, m;
+    setupScene() {
 
-        var renderOrModel = function (node) {
-            return node.render ? node.render : node.model;
-        };
+        // lightmapper needs original model draw calls
+        this.revertStatic = false;
+        if (this.scene._needsStaticPrepare) {
+            this.scene._needsStaticPrepare = false;
+            this.revertStatic = true;
+        }
 
-        var renderOrModelMeshInstances = function (node) {
-            return node.render ? node.render.meshInstances : node.model.model.meshInstances;
-        };
+        // backup
+        this.fog = this.scene.fog;
+        this.ambientLight = this.scene.ambientLight.clone();
```

Comment:
Is this not allocated/defined in the Lightmapper constructor? Ideally, you'd create the `Color` there, and here, you'd just copy.

Reply:
done, thanks!